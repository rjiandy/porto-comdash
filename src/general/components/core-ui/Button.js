// @flow

import React from 'react';
import MaterialIcon, {Icon} from './Icon';
import {FlatButton, RaisedButton} from 'material-ui';
import {THEME_COLOR, DARKER_BLUE} from '../../constants/colors';
import {ALTERNATIVE_FONT_FAMILY, SMALL_FONT_SIZE} from '../../constants/text';
import wrapComponent from '../../helpers/wrapComponent';

type Props = {
  label?: string;
  labelStyle?: {[key: string]: string | number};
  onPress: () => void;
  raised?: boolean;
  primary?: boolean;
  secondary?: boolean;
  disabled?: boolean;
  fullWidth?: boolean;
  style?: {[key: string]: string | number};
  icon?: string;
  iconSvg?: string;
  svgContainerStyle?: {[key: string]: string | number};
  iconStyle?: {[key: string]: string | number};
  iconPosition?: 'left' | 'right';
};

const DEFAULT_PRIMARY_BUTTON_PROPS = {
  backgroundColor: THEME_COLOR,
  hoverColor: DARKER_BLUE,
};

const DEFAULT_SECONDARY_BUTTON_PROPS = {
  backgroundColor: 'white',
};

export function Button(props: Props) {
  let {
    onPress,
    icon,
    iconSvg,
    svgContainerStyle,
    primary,
    secondary,
    iconPosition,
    iconStyle,
    raised,
    style,
    labelStyle,
    disabled,
    ...otherProps
  } = props;

  let defaultProps = {};
  let composedButtonStyle = {};
  let composedLabelStyle = {};
  let composedIconStyle = {};

  if (primary) {
    Object.assign(defaultProps, DEFAULT_PRIMARY_BUTTON_PROPS);
    Object.assign(composedButtonStyle, styles.primaryButton);
    Object.assign(
      composedLabelStyle,
      styles.primaryLabel,
      styles.font,
      labelStyle,
    );
    Object.assign(composedIconStyle, styles.primaryIcon, iconStyle);
  } else if (secondary) {
    Object.assign(defaultProps, DEFAULT_SECONDARY_BUTTON_PROPS);
    Object.assign(composedButtonStyle, styles.secondaryButton);
    Object.assign(
      composedLabelStyle,
      styles.secondaryLabel,
      styles.font,
      labelStyle,
    );
    Object.assign(composedIconStyle, styles.secondaryIcon, iconStyle);
  } else {
    Object.assign(composedButtonStyle, styles.defaultButton);
    Object.assign(composedLabelStyle, styles.font, labelStyle);
    Object.assign(composedIconStyle, iconStyle);
  }

  let ButtonComponent;
  let buttonStyle = {};
  if (raised) {
    ButtonComponent = RaisedButton;
    buttonStyle = {
      disabledBackgroundColor: 'grey',
      style: Object.assign({}, styles.defaultButton, style),
      buttonStyle: {
        ...composedButtonStyle,
        cursor: disabled && 'not-allowed',
      },
      labelStyle: composedLabelStyle,
    };
  } else {
    ButtonComponent = FlatButton;
    buttonStyle = {
      style: {
        cursor: disabled && 'not-allowed',
        ...composedButtonStyle,
        ...style,
      },
      labelStyle: composedLabelStyle,
    };
  }

  let iconComponent;
  let iconPositionInButton;

  if (icon) {
    iconComponent = <MaterialIcon name={icon} style={composedIconStyle} />;
  }
  if (iconSvg) {
    iconComponent = (
      <Icon
        name={iconSvg}
        style={composedIconStyle}
        containerStyle={svgContainerStyle}
      />
    );
  }
  if (iconPosition && iconPosition === 'right') {
    iconPositionInButton = {labelPosition: 'before'};
  } else {
    iconPositionInButton = {labelPosition: 'after'};
  }
  // TODO set up background color for disabled
  return (
    <ButtonComponent
      onTouchTap={onPress}
      icon={iconComponent}
      backgroundColor={disabled && 'grey'}
      disabled={disabled}
      {...iconPositionInButton}
      {...buttonStyle}
      {...defaultProps}
      {...otherProps}
    />
  );
}

const styles = {
  defaultButton: {
    borderRadius: 3,
  },
  primaryButton: {
    borderRadius: 3,
    flexDirection: 'row',
  },
  secondaryButton: {
    border: `1px solid ${THEME_COLOR}`,
    borderRadius: 3,
    flexDirection: 'row',
  },
  font: {
    fontFamily: ALTERNATIVE_FONT_FAMILY,
    fontSize: SMALL_FONT_SIZE,
    fontWeight: '200',
  },
  primaryLabel: {
    color: 'white',
  },
  secondaryLabel: {
    color: THEME_COLOR,
  },
  primaryIcon: {
    color: 'white',
  },
  secondaryIcon: {
    color: THEME_COLOR,
  },
};

export default wrapComponent(Button);
