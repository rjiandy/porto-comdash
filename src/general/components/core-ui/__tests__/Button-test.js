import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';

import {Button} from '../Button';

it('should render button', () => {
  let onPress = jest.fn();
  let button = shallow(<Button label="click me" onPress={onPress} />);
  expect(toJSON(button)).toMatchSnapshot();
  button.simulate('touchTap');
  expect(onPress).toBeCalled();
});

it('should render button with type according to props', () => {
  let onPress = jest.fn();
  let button = shallow(
    <Button label="click me" onPress={onPress} secondary={true} />
  );
  expect(toJSON(button).props.style).toMatchSnapshot();
});

it('should render button with label after or before the icon according to props', () => {
  let onPress = jest.fn();
  let button = shallow(
    <Button
      label="click me"
      onPress={onPress}
      icon="mail"
      iconPosition="right"
    />
  );
  expect(button.prop('labelPosition')).toEqual('before');

  button = shallow(
    <Button
      label="click me"
      onPress={onPress}
      icon="mail"
      iconPosition="left"
    />
  );
  expect(button.prop('labelPosition')).toEqual('after');

  button = shallow(
    <Button
      onPress={onPress}
      icon="mail"
    />
  );
  expect(button.prop('labelPosition')).toEqual('after');
});
