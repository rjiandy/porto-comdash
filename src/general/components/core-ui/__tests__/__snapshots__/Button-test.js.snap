// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render button 1`] = `
<FlatButton
  disabled={false}
  fullWidth={false}
  label="click me"
  labelPosition="after"
  labelStyle={
    Object {
      "fontFamily": "SanFransiscoDisplay",
      "fontSize": 11,
      "fontWeight": "200",
    }
  }
  onKeyboardFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchStart={[Function]}
  onTouchTap={[Function]}
  primary={false}
  secondary={false}
  style={
    Object {
      "borderRadius": 3,
      "cursor": undefined,
    }
  }
/>
`;

exports[`should render button with type according to props 1`] = `
Object {
  "border": "1px solid #4A90E2",
  "borderRadius": 3,
  "cursor": undefined,
  "flexDirection": "row",
}
`;
